File: packages/runtime-go/a2ui.go
package runtime

import (
	"encoding/json"
	"fmt"
	"strings"
)

const a2uiSchemaDescription = "A2UI Component Schema — available components for generating UI surfaces. Use these component names and properties when creating A2UI operations."
const basicCatalog = "https://a2ui.org/specification/v0_9/basic_catalog.json"

// A2UIConfig configures native v0.9 surface middleware. Retries remain agent-owned.
type A2UIConfig struct {
	Enabled          *bool    `json:"enabled,omitempty"`
	Agents           []string `json:"agents,omitempty"`
	InjectA2UITool   any      `json:"injectA2UITool,omitempty"`
	DefaultCatalogID string   `json:"defaultCatalogId,omitempty"`
	Schema           any      `json:"schema,omitempty"`
	ToolNames        []string `json:"a2uiToolNames,omitempty"`
}

func (c *A2UIConfig) enabled(agent string) bool {
	if c == nil || (c.Enabled != nil && !*c.Enabled) {
		return false
	}
	if len(c.Agents) == 0 {
		return true
	}
	for _, id := range c.Agents {
		if id == agent {
			return true
		}
	}
	return false
}

// resolveA2UI copies request defaults so one frontend catalog cannot affect another request.
func resolveA2UI(config *A2UIConfig, agent string, hasCatalog bool) *A2UIConfig {
	if config == nil && !hasCatalog {
		return nil
	}
	copy := A2UIConfig{}
	if config != nil {
		copy = *config
	}
	if !copy.enabled(agent) {
		return nil
	}
	if copy.InjectA2UITool == nil && hasCatalog {
		copy.InjectA2UITool = true
	}
	return &copy
}
func (c *A2UIConfig) toolName() string {
	if name := str(c.InjectA2UITool); name != "" {
		return name
	}
	return "render_a2ui"
}
func (c *A2UIConfig) isTool(name string) bool {
	if len(c.ToolNames) == 0 && name == "render_a2ui" {
		return true
	}
	for _, n := range c.ToolNames {
		if name == n {
			return true
		}
	}
	return c.InjectA2UITool != nil && c.InjectA2UITool != false && name == c.toolName()
}
func processA2UIAction(input map[string]any) {
	action := object(object(object(input["forwardedProps"])["a2uiAction"])["userAction"])
	if len(action) == 0 {
		return
	}
	call := uuid()
	args, _ := json.Marshal(action)
	messages, _ := input["messages"].([]any)
	name, surface := str(action["name"]), str(action["surfaceId"])
	if name == "" {
		name = "unknown_action"
	}
	if surface == "" {
		surface = "unknown_surface"
	}
	text := fmt.Sprintf("User performed action %q on surface %q", name, surface)
	if component := str(action["sourceComponentId"]); component != "" {
		text += fmt.Sprintf(" (component: %s)", component)
	}
	context, _ := json.Marshal(object(action["context"]))
	text += ". Context: " + string(context)
	input["messages"] = append(messages, map[string]any{"id": uuid(), "role": "assistant", "content": "", "toolCalls": []any{map[string]any{"id": call, "type": "function", "function": map[string]any{"name": "log_a2ui_event", "arguments": string(args)}}}}, map[string]any{"id": uuid(), "role": "tool", "toolCallId": call, "content": text})
}
func prepareA2UI(c *A2UIConfig, input map[string]any) string {
	context, _ := input["context"].([]any)
	catalog := ""
	for _, entry := range context {
		m := object(entry)
		if m["description"] == a2uiSchemaDescription {
			var schema map[string]any
			json.Unmarshal([]byte(str(m["value"])), &schema)
			catalog = str(schema["catalogId"])
		}
	}
	if c.Schema != nil {
		filtered := []any{}
		for _, entry := range context {
			if object(entry)["description"] != a2uiSchemaDescription {
				filtered = append(filtered, entry)
			}
		}
		raw, _ := json.Marshal(c.Schema)
		context = append(filtered, map[string]any{"description": a2uiSchemaDescription, "value": string(raw)})
	}
	if c.InjectA2UITool != nil && c.InjectA2UITool != false {
		tools, _ := input["tools"].([]any)
		filtered := []any{}
		for _, tool := range tools {
			if object(tool)["name"] != c.toolName() {
				filtered = append(filtered, tool)
			}
		}
		input["tools"] = append(filtered, map[string]any{"name": c.toolName(), "description": "Render a dynamic A2UI v0.9 surface with structured parameters.", "parameters": map[string]any{"type": "object", "properties": map[string]any{"surfaceId": map[string]any{"type": "string"}, "components": map[string]any{"type": "array", "items": map[string]any{"type": "object"}}, "data": map[string]any{"type": "object"}}, "required": []any{"surfaceId", "components"}}})
		forwarded := object(input["forwardedProps"])
		forwarded["injectA2UITool"] = c.InjectA2UITool
		input["forwardedProps"] = forwarded
		context = append(context, map[string]any{"description": "A2UI render tool usage guide — how to call " + c.toolName() + " with valid arguments.", "value": "Use A2UI v0.9 flat components with unique IDs and root id root. Supply surfaceId and components. Use catalog component types and required properties. Reference child IDs; do not nest components or create cycles. Data bindings use {path: /key}; repeated children use {componentId: id, path: /items}. Supply data for bindings."})
	}
	input["context"] = context
	return catalog
}

type a2uiCall struct {
	key, args, surface, catalog, dataKey string
	components                           []any
	rejected, painted, complete, result  bool
	items                                int
}
type a2uiStream struct {
	config          *A2UIConfig
	frontend, outer string
	calls           map[string]*a2uiCall
	painted         map[string]bool
	order           []string
	attempts        map[string]int
}

func newA2UI(c *A2UIConfig, frontend string) *a2uiStream {
	return &a2uiStream{config: c, frontend: frontend, calls: map[string]*a2uiCall{}, painted: map[string]bool{}, attempts: map[string]int{}}
}
func (s *a2uiStream) activity(key string, content map[string]any) Event {
	return Event{"type": "ACTIVITY_SNAPSHOT", "messageId": "a2ui-surface-" + key, "activityType": "a2ui-surface", "content": content, "replace": true}
}
func (s *a2uiStream) process(e Event) []Event {
	id := str(e["toolCallId"])
	events := []Event{}
	switch str(e["type"]) {
	case "TOOL_CALL_START":
		if s.config.isTool(str(e["toolCallName"])) {
			key := id
			if s.outer != "" {
				key = s.outer
			}
			s.attempts[key]++
			s.calls[id] = &a2uiCall{key: key, dataKey: "items"}
			s.order = append(s.order, id)
			events = append(events, s.activity(key, map[string]any{"status": "building"}))
		} else if e["toolCallName"] != "log_a2ui_event" {
			s.outer = id
		}
	case "TOOL_CALL_ARGS":
		if call := s.calls[id]; call != nil && !call.rejected {
			call.args += str(e["delta"])
			if len(call.args) > 4<<20 {
				call.rejected = true
				events = append(events, s.activity(call.key, map[string]any{"status": "failed", "error": "A2UI arguments exceed size limit"}))
				break
			}
			if !call.painted {
				call.surface = fieldString(call.args, "surfaceId")
				if call.surface == "" {
					call.surface = "surface-" + id
				}
				call.catalog = s.config.DefaultCatalogID
				if call.catalog == "" {
					call.catalog = s.frontend
				}
				if call.catalog == "" {
					call.catalog = fieldString(call.args, "catalogId")
				}
				if call.catalog == "" || call.catalog == "basic" {
					call.catalog = basicCatalog
				}
			}
			advanced := false
			if !call.painted {
				raw := fieldRemainder(call.args, "components")
				var components []any
				if json.Unmarshal([]byte(raw), &components) == nil {
					validation := validateComponents(components, object(s.config.Schema))
					if len(validation) > 0 {
						call.rejected = true
						events = append(events, s.activity(call.key, map[string]any{"status": "retrying", "attempt": s.attempts[call.key] + 1, "maxAttempts": 3, "errors": validation}))
						break
					}
					call.components = components
					call.painted = true
					advanced = true
					s.painted[call.surface] = true
					for _, component := range components {
						path := str(object(object(component)["children"])["path"])
						if path != "" {
							call.dataKey = strings.TrimPrefix(path, "/")
							break
						}
					}
				}
			}
			if call.painted {
				var data any
				rawData := fieldRemainder(call.args, "data")
				items, _ := partialArray(fieldRemainder(rawData, call.dataKey))
				if len(items) > call.items {
					call.items = len(items)
					advanced = true
					data = map[string]any{call.dataKey: items}
				}
				var complete map[string]any
				if !call.complete && json.Unmarshal([]byte(rawData), &complete) == nil {
					call.complete = true
					data = complete
					advanced = true
				}
				if advanced {
					events = append(events, s.snapshot(call, data))
				}
			}
		}
	case "TOOL_CALL_RESULT":
		if call := s.calls[id]; call != nil {
			call.result = true
		}
		var result map[string]any
		if json.Unmarshal([]byte(str(e["content"])), &result) == nil {
			if result["code"] == "a2ui_recovery_exhausted" {
				key := id
				if s.outer != "" {
					key = s.outer
				}
				events = append(events, s.activity(key, map[string]any{"status": "failed", "error": result["error"], "attempts": result["attempts"], "maxAttempts": 3}))
			} else if operations, ok := result["a2ui_operations"].([]any); ok {
				groups := map[string][]any{}
				order := []string{}
				for _, op := range operations {
					m := object(op)
					surface := ""
					for _, kind := range []string{"createSurface", "updateComponents", "updateDataModel", "deleteSurface"} {
						if value, ok := m[kind]; ok {
							surface = str(object(value)["surfaceId"])
						}
					}
					if surface == "" || s.painted[surface] {
						continue
					}
					if _, ok := groups[surface]; !ok {
						order = append(order, surface)
					}
					groups[surface] = append(groups[surface], op)
				}
				for _, surface := range order {
					key := id
					if s.outer != "" {
						key = s.outer
					}
					if len(order) > 1 {
						key = surface + "-" + key
					}
					events = append(events, s.activity(key, map[string]any{"a2ui_operations": groups[surface]}))
				}
			}
		}
		if s.outer == id {
			s.outer = ""
		}
	}
	return events
}
func (s *a2uiStream) snapshot(call *a2uiCall, data any) Event {
	ops := []any{map[string]any{"version": "v0.9", "createSurface": map[string]any{"surfaceId": call.surface, "catalogId": call.catalog}}, map[string]any{"version": "v0.9", "updateComponents": map[string]any{"surfaceId": call.surface, "components": call.components}}}
	if data != nil {
		ops = append(ops, map[string]any{"version": "v0.9", "updateDataModel": map[string]any{"surfaceId": call.surface, "path": "/", "value": data}})
	}
	return s.activity(call.key, map[string]any{"a2ui_operations": ops})
}
func (s *a2uiStream) finish() []Event {
	events := []Event{}
	for _, id := range s.order {
		call := s.calls[id]
		if !call.result {
			content := `{"status":"rendered"}`
			if !call.painted {
				content = `{"status":"error","error":"A2UI surface was not valid"}`
			}
			events = append(events, Event{"type": "TOOL_CALL_RESULT", "messageId": uuid(), "toolCallId": id, "content": content})
		}
	}
	return events
}

// fieldRemainder locates a root member with a real JSON decoder, including unfinished values.
func fieldRemainder(raw, key string) string {
	d := json.NewDecoder(strings.NewReader(raw))
	token, e := d.Token()
	if e != nil || token != json.Delim('{') {
		return ""
	}
	for d.More() {
		token, e = d.Token()
		if e != nil {
			return ""
		}
		start := int(d.InputOffset())
		for start < len(raw) && (raw[start] == ':' || raw[start] == ' ' || raw[start] == '\n' || raw[start] == '\r' || raw[start] == '\t') {
			start++
		}
		var value json.RawMessage
		e = d.Decode(&value)
		if token == key {
			if e == nil {
				return string(value)
			}
			return raw[start:]
		}
		if e != nil {
			return ""
		}
	}
	return ""
}
func fieldString(raw, key string) string {
	var value string
	json.Unmarshal([]byte(fieldRemainder(raw, key)), &value)
	return value
}
func partialArray(raw string) ([]any, bool) {
	d := json.NewDecoder(strings.NewReader(raw))
	token, e := d.Token()
	if e != nil || token != json.Delim('[') {
		return nil, false
	}
	items := []any{}
	for d.More() {
		var item any
		if d.Decode(&item) != nil {
			return items, false
		}
		items = append(items, item)
	}
	token, e = d.Token()
	return items, e == nil && token == json.Delim(']')
}

File: packages/runtime-python/src/copilotkit_runtime/a2ui.py
"""Native A2UI v0.9 middleware, aligned with AG-UI middleware 0.0.10."""

import copy
import json
from collections.abc import AsyncIterator
from dataclasses import dataclass
from typing import Any
from uuid import uuid4

from .models import Json

SCHEMA_DESCRIPTION = "A2UI Component Schema — available components for generating UI surfaces. Use these component names and properties when creating A2UI operations."
BASIC_CATALOG = "https://a2ui.org/specification/v0_9/basic_catalog.json"
_DECODER = json.JSONDecoder()
_OPS = ("createSurface", "updateComponents", "updateDataModel", "deleteSurface")


@dataclass(frozen=True)
class A2UIConfig:
    """Configure per-agent A2UI context, tool injection and recovery presentation."""

    enabled: bool = True
    agents: tuple[str, ...] | None = None
    schema: Json | list[Json] | None = None
    inject_tool: bool | str = False
    tool_names: tuple[str, ...] = ("render_a2ui",)
    default_catalog_id: str | None = None
    max_attempts: int = 3
    show_progress_tokens: bool = True
    debug_exposure: str | None = None
    max_argument_bytes: int = 2 * 1024 * 1024

    def __post_init__(self) -> None:
        """Validate public configuration before a run starts."""
        if not isinstance(self.inject_tool, (bool, str)) or self.inject_tool == "":
            raise ValueError("inject_tool must be a bool or nonempty tool name")
        if self.max_attempts < 1 or self.max_argument_bytes < 1:
            raise ValueError("A2UI limits must be positive")
        if self.debug_exposure not in (None, "hidden", "collapsed", "verbose"):
            raise ValueError("Invalid A2UI debug exposure")

    def applies(self, agent_id: str) -> bool:
        """Return whether A2UI is enabled for this configured agent."""
        return self.enabled and (self.agents is None or agent_id in self.agents)


def _references(value: Any, path: str) -> list[tuple[str, str]]:
    """Extract references from a string, template or reference list."""
    if isinstance(value, str):
        return [(path, value)]
    if isinstance(value, dict) and isinstance(value.get("componentId"), str):
        return [(path, value["componentId"])]
    if isinstance(value, list):
        return [
            edge
            for index, item in enumerate(value)
            for edge in _references(item, f"{path}[{index}]")
        ]
    return []


def _edges(component: Json, schema: Json) -> list[tuple[str, str]]:
    """Read implicit children plus catalog-marked component references."""
    result = _references(component.get("child"), "child") + _references(
        component.get("children"), "children"
    )
    for field, prop in schema.get("properties", {}).items():
        if field in ("child", "children") or not isinstance(prop, dict):
            continue
        if prop.get("format") in ("componentRef", "componentRefList"):
            result.extend(_references(component.get(field), field))
        elif prop.get("type") == "array" and isinstance(prop.get("items"), dict):
            values = component.get(field)
            if not isinstance(values, list):
                continue
            for index, item in enumerate(values):
                if not isinstance(item, dict):
                    continue
                for name, sub in prop["items"].get("properties", {}).items():
                    if isinstance(sub, dict) and sub.get("format") in (
                        "componentRef",
                        "componentRefList",
                    ):
                        result.extend(_references(item.get(name), f"{field}[{index}].{name}"))
    return result


def validate_components(components: Any, catalog: Json | None = None) -> list[Json]:
    """Validate the atomic component tree before paint; bindings remain deferred."""
    errors: list[Json] = []

    def add(code: str, path: str, message: str) -> None:
        errors.append({"code": code, "path": path, "message": message})

    if not isinstance(components, list) or not components:
        add("empty_components", "components", "A2UI components must be a non-empty array")
        return errors
    ids: set[str] = set()
    adjacency: dict[str, list[str]] = {}
    for component in components:
        identifier = component.get("id") if isinstance(component, dict) else None
        if isinstance(identifier, str):
            if identifier in ids:
                add(
                    "duplicate_id",
                    f"components[id={identifier}]",
                    f"Duplicate component id '{identifier}'",
                )
            ids.add(identifier)
    for index, component in enumerate(components):
        component = component if isinstance(component, dict) else {}
        identifier, kind = component.get("id"), component.get("component")
        if not isinstance(identifier, str) or not identifier:
            add(
                "missing_id",
                f"components[{index}].id",
                f"Component at index {index} is missing a string 'id'",
            )
        if not isinstance(kind, str) or not kind:
            add(
                "missing_component_type",
                f"components[{index}].component",
                f"Component at index {index} is missing a string 'component' type",
            )
        schema = (
            (catalog or {}).get("components", {}).get(kind, {}) if isinstance(kind, str) else {}
        )
        if catalog and isinstance(kind, str):
            if kind not in catalog.get("components", {}):
                add(
                    "unknown_component",
                    f"components[{index}].component",
                    f"Component type '{kind}' is not in the catalog",
                )
            else:
                for prop in schema.get("required", []):
                    if prop not in component:
                        add(
                            "missing_required_prop",
                            f"components[{index}].{prop}",
                            f"Component '{kind}' (index {index}) is missing required prop '{prop}'",
                        )
        edges = _edges(component, schema)
        for path, ref in edges:
            if ref not in ids:
                add(
                    "unresolved_child",
                    f"components[{index}].{path}",
                    f"Child reference '{ref}' does not match any component id",
                )
        if isinstance(identifier, str):
            adjacency[identifier] = [ref for _, ref in edges]
    colors: dict[str, int] = {}
    cycles: set[tuple[str, ...]] = set()
    for root in adjacency:
        if colors.get(root):
            continue
        stack = [(root, 0)]
        chain = [root]
        colors[root] = 1
        while stack:
            node, index = stack[-1]
            neighbors = adjacency.get(node, [])
            if index >= len(neighbors):
                colors[node] = 2
                stack.pop()
                chain.pop()
                continue
            stack[-1] = (node, index + 1)
            child = neighbors[index]
            if not colors.get(child):
                colors[child] = 1
                stack.append((child, 0))
                chain.append(child)
            elif colors[child] == 1:
                cycle = chain[chain.index(child) :]
                smallest = cycle.index(min(cycle))
                canonical = tuple(cycle[smallest:] + cycle[:smallest])
                if canonical not in cycles:
                    cycles.add(canonical)
                    add(
                        "child_cycle",
                        f"components[id={canonical[0]}]",
                        "Child reference cycle detected: "
                        + " -> ".join([*canonical, canonical[0]]),
                    )
    if "root" not in ids:
        add("no_root", "components", "No component has id 'root'")
    return errors


def _field(raw: str, name: str) -> str | None:
    """Find a root property without mistaking quoted content for a field name."""
    raw = raw.lstrip()
    if not raw.startswith("{"):
        return None
    position = 1
    try:
        while position < len(raw):
            position += len(raw[position:]) - len(raw[position:].lstrip())
            key, end = _DECODER.raw_decode(raw, position)
            position = end
            position += len(raw[position:]) - len(raw[position:].lstrip())
            if raw[position] != ":":
                return None
            position += 1
            position += len(raw[position:]) - len(raw[position:].lstrip())
            if key == name:
                return raw[position:]
            _, position = _DECODER.raw_decode(raw, position)
            position += len(raw[position:]) - len(raw[position:].lstrip())
            if raw[position] != ",":
                return None
            position += 1
    except (ValueError, IndexError):
        return None
    return None


def _value(raw: str, field: str) -> Any:
    """Read a complete root value from otherwise partial JSON."""
    fragment = _field(raw, field)
    if fragment is None:
        return None
    try:
        return _DECODER.raw_decode(fragment)[0]
    except ValueError:
        return None


def _array(fragment: str | None) -> tuple[list[Any], bool]:
    """Return only complete array entries, plus the array closure status."""
    if fragment is None or not fragment.startswith("["):
        return [], False
    items: list[Any] = []
    position = 1
    try:
        while position < len(fragment):
            position += len(fragment[position:]) - len(fragment[position:].lstrip())
            if fragment[position] == "]":
                return items, True
            item, position = _DECODER.raw_decode(fragment, position)
            items.append(item)
            position += len(fragment[position:]) - len(fragment[position:].lstrip())
            if fragment[position] == "]":
                return items, True
            if fragment[position] != ",":
                return items, False
            position += 1
    except (ValueError, IndexError):
        pass
    return items, False


class A2UIMiddleware:
    """Create isolated stream state per run; model retry decisions stay with agents."""

    def __init__(self, config: A2UIConfig) -> None:
        self.config = config
        self.names = set(config.tool_names)
        self.tool_name = (
            config.inject_tool if isinstance(config.inject_tool, str) else "render_a2ui"
        )
        if config.inject_tool:
            self.names.add(self.tool_name)

    def prepare(self, input: Json) -> Json:
        """Inject server context, optional tool schema, and synthetic action history."""
        result = copy.deepcopy(input)
        context = result.setdefault("context", [])
        forwarded = result.setdefault("forwardedProps", {})
        action = forwarded.get("a2uiAction", {}).get("userAction")
        if isinstance(action, dict):
            call_id = str(uuid4())
            text = f'User performed action "{action.get("name", "unknown_action")}" on surface "{action.get("surfaceId", "unknown_surface")}"'
            if action.get("sourceComponentId"):
                text += f" (component: {action['sourceComponentId']})"
            text += ". Context: " + json.dumps(action.get("context", {}), separators=(",", ":"))
            result.setdefault("messages", []).extend(
                [
                    {
                        "id": str(uuid4()),
                        "role": "assistant",
                        "content": "",
                        "toolCalls": [
                            {
                                "id": call_id,
                                "type": "function",
                                "function": {
                                    "name": "log_a2ui_event",
                                    "arguments": json.dumps(action),
                                },
                            }
                        ],
                    },
                    {"id": str(uuid4()), "role": "tool", "toolCallId": call_id, "content": text},
                ]
            )
        if self.config.schema:
            context[:] = [
                entry for entry in context if entry.get("description") != SCHEMA_DESCRIPTION
            ]
            context.append(
                {"description": SCHEMA_DESCRIPTION, "value": json.dumps(self.config.schema)}
            )
        if self.config.inject_tool:
            result["tools"] = [
                tool for tool in result.get("tools", []) if tool.get("name") != self.tool_name
            ]
            result["tools"].append(
                {
                    "name": self.tool_name,
                    "description": "Render a dynamic A2UI v0.9 surface with structured parameters. Follow the A2UI render tool usage guide provided in context.",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "surfaceId": {
                                "type": "string",
                                "description": "Unique surface identifier.",
                            },
                            "components": {"type": "array", "items": {"type": "object"}},
                            "data": {"type": "object"},
                        },
                        "required": ["surfaceId", "components"],
                    },
                }
            )
            forwarded["injectA2UITool"] = self.config.inject_tool
            description = (
                f"A2UI render tool usage guide — how to call {self.tool_name} with valid arguments."
            )
            context[:] = [entry for entry in context if entry.get("description") != description]
            context.append(
                {
                    "description": description,
                    "value": f'Call {self.tool_name} with surfaceId and a nonempty components array. Use flat v0.9 components with unique id and component fields. Include id root. Reference children by ID. Only use catalog component names and required properties. Data bindings use {{"path":"/key"}}. Repeated children use {{"componentId":"item","path":"/items"}}. Supply data for bindings. The host owns catalogId. Actions use an event object with name and context. Do not invent image URLs.',
                }
            )
        return result

    def _activity(self, key: str, content: Json) -> Json:
        """Keep building/retry/failure/paint states on one replaceable message ID."""
        if "status" in content and self.config.debug_exposure:
            content = {**content, "debugExposure": self.config.debug_exposure}
        return {
            "type": "ACTIVITY_SNAPSHOT",
            "messageId": "a2ui-surface-" + key,
            "activityType": "a2ui-surface",
            "content": content,
            "replace": True,
        }

    async def transform(self, source: AsyncIterator[Json], original: Json) -> AsyncIterator[Json]:
        """Emit validated atomic components and cumulative progressive data before persistence."""
        calls: dict[str, Json] = {}
        resolved = {
            message.get("toolCallId")
            for message in original.get("messages", [])
            if message.get("role") == "tool"
        }
        outer: str | None = None
        painted: set[str] = set()
        retries: set[str] = set()
        attempts: dict[str, int] = {}
        frontend_catalog = None
        for entry in original.get("context", []):
            if entry.get("description") == SCHEMA_DESCRIPTION:
                try:
                    frontend_catalog = json.loads(entry["value"]).get("catalogId")
                except (ValueError, AttributeError, KeyError):
                    pass
        catalog = (
            self.config.schema
            if isinstance(self.config.schema, dict) and self.config.schema.get("components")
            else None
        )
        async for event in source:
            kind, call_id = event.get("type"), event.get("toolCallId", "")
            if kind == "TOOL_CALL_START":
                if event.get("toolCallName") in self.names:
                    key = outer or call_id
                    attempts[key] = attempts.get(key, 0) + 1
                    calls[call_id] = {
                        "args": "",
                        "key": key,
                        "components": None,
                        "rejected": False,
                        "items": 0,
                        "complete": False,
                        "tokens": 0,
                    }
                    if key not in retries:
                        yield self._activity(key, {"status": "building"})
                elif event.get("toolCallName") != "log_a2ui_event":
                    outer = call_id
            elif kind == "TOOL_CALL_ARGS" and call_id in calls:
                state = calls[call_id]
                state["args"] += event.get("delta", "")
                if len(state["args"].encode()) > self.config.max_argument_bytes:
                    raise ValueError("A2UI arguments exceed configured limit")
                raw, key = state["args"], state["key"]
                tokens = int(len(raw) / 4 + 0.5)
                if (
                    self.config.show_progress_tokens
                    and not state["components"]
                    and key not in retries
                    and tokens - state["tokens"] >= 20
                ):
                    state["tokens"] = tokens
                    yield self._activity(key, {"status": "building", "progressTokens": tokens})
                surface = _value(raw, "surfaceId") or call_id
                streamed_catalog = _value(raw, "catalogId")
                catalog_id = (
                    self.config.default_catalog_id
                    or frontend_catalog
                    or (
                        streamed_catalog
                        if streamed_catalog and streamed_catalog != "basic"
                        else BASIC_CATALOG
                    )
                )
                components_advanced = False
                if state["components"] is None and not state["rejected"]:
                    components, closed = _array(_field(raw, "components"))
                    if closed:
                        errors = validate_components(components, catalog)
                        if errors:
                            state["rejected"] = True
                            retries.add(key)
                            yield self._activity(
                                key,
                                {
                                    "status": "retrying",
                                    "attempt": min(attempts[key] + 1, self.config.max_attempts),
                                    "maxAttempts": self.config.max_attempts,
                                    "errors": errors,
                                },
                            )
                        else:
                            state["components"] = components
                            components_advanced = True
                            state["surface"] = surface
                            state["catalog"] = catalog_id
                if state["components"]:
                    surface = state["surface"]
                    catalog_id = state["catalog"]
                    data_key = next(
                        (
                            component["children"]["path"].removeprefix("/")
                            for component in state["components"]
                            if isinstance(component.get("children"), dict)
                            and isinstance(component["children"].get("path"), str)
                        ),
                        "items",
                    )
                    data_fragment = _field(raw, "data")
                    items, _ = _array(_field(data_fragment, data_key) if data_fragment else None)
                    ops = [
                        {
                            "version": "v0.9",
                            "createSurface": {"surfaceId": surface, "catalogId": catalog_id},
                        },
                        {
                            "version": "v0.9",
                            "updateComponents": {
                                "surfaceId": surface,
                                "components": state["components"],
                            },
                        },
                    ]
                    if components_advanced or (
                        len(items) > state["items"] and not state["complete"]
                    ):
                        state["items"] = len(items)
                        progressive = ops + (
                            [
                                {
                                    "version": "v0.9",
                                    "updateDataModel": {
                                        "surfaceId": surface,
                                        "path": "/",
                                        "value": {data_key: items},
                                    },
                                }
                            ]
                            if items
                            else []
                        )
                        yield self._activity(key, {"a2ui_operations": progressive})
                        painted.add(surface)
                        retries.discard(key)
                    data = _value(raw, "data")
                    if isinstance(data, dict) and not state["complete"]:
                        state["complete"] = True
                        yield self._activity(
                            key,
                            {
                                "a2ui_operations": ops
                                + [
                                    {
                                        "version": "v0.9",
                                        "updateDataModel": {
                                            "surfaceId": surface,
                                            "path": "/",
                                            "value": data,
                                        },
                                    }
                                ]
                            },
                        )
            if kind == "RUN_FINISHED":
                for pending_id in calls.keys() - resolved:
                    yield {
                        "type": "TOOL_CALL_RESULT",
                        "messageId": str(uuid4()),
                        "toolCallId": pending_id,
                        "content": json.dumps({"status": "rendered"}),
                    }
                    resolved.add(pending_id)
            yield event
            if kind == "TOOL_CALL_RESULT":
                resolved.add(call_id)
                try:
                    content = json.loads(event.get("content", ""))
                except (ValueError, TypeError):
                    content = None
                if isinstance(content, dict):
                    operations = content.get("a2ui_operations")
                    if isinstance(operations, list):
                        groups: dict[str, list[Json]] = {}
                        for operation in operations:
                            if not isinstance(operation, dict):
                                continue
                            op_name = next(
                                (name for name in _OPS if isinstance(operation.get(name), dict)),
                                None,
                            )
                            if op_name:
                                surface = operation[op_name].get("surfaceId", "default")
                                if surface not in painted:
                                    groups.setdefault(surface, []).append(operation)
                        for surface, group in groups.items():
                            key = outer or call_id
                            if len(groups) > 1:
                                key = surface + "-" + key
                            yield self._activity(key, {"a2ui_operations": group})
                    elif content.get("code") == "a2ui_recovery_exhausted":
                        history = content.get("attempts", [])
                        yield self._activity(
                            outer or call_id,
                            {
                                "status": "failed",
                                "error": content.get("error", "A2UI generation failed"),
                                "attempts": history,
                                "maxAttempts": len(history) or self.config.max_attempts,
                            },
                        )
                if call_id == outer:
                    outer = None
